File: Revision_Spring2017/AdaptiveFilter.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% Author: David Fridovich-Keil
% File: ConstantVelocityKF.m
%
% Implements a linear adaptive filter with LMS single-point expectation
% approximation.
% 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

classdef AdaptiveFilter < handle
    properties
        a_; % Autoregressive coefficients (on previous latencies).
        b_; % Moving average coefficients (on current/previous controls).
        x_; % Relevant history of previous latencies. Java ArrayList.
        u_; % Relevant history of previous controls. Java ArrayList.
        alpha_; % Learning rate for 'a'.
        beta_; % Learning rate for 'b'.
    end
    
    methods
        % Constructor. Initialize filters with i.i.d. Gaussian.
        % Must provide the following inputs:
        % 'p' = number of autoregressive coefficients
        % 'q' = number of moving average coefficients
        % 'alpha' = learning rate for 'a'
        % 'beta' = learning rate for 'b'
        function lms = AdaptiveFilter(p, q, alpha, beta)
            import java.util.ArrayList
            
            lms.a_ = 0.1 * randn(p, 1);
            lms.b_ = 0.1 * randn(q, 1);
            lms.x_ = ArrayList();
            lms.u_ = ArrayList();
            lms.alpha_ = alpha;
            lms.beta_ = beta;
        end
        
        % Predict using the current coefficients. Must supply a control
        % input 'u'.
        function x_hat = Predict(lms, u)
            x_hat = 0;
            
            % Add 'u' to the control history.
            lms.u_.add(u);

            % Get sizes of histories.
            x_size = lms.x_.size();
            u_size = lms.u_.size();
            
            % Convolve 'a' with latency history.
            for ii = 1:min(x_size, numel(lms.a_))
                x_index = x_size - ii;
                x_hat = x_hat + lms.x_.get(x_index) * lms.a_(ii);
            end
            
            % Convolve 'b' with control history.
            for ii = 1:min(u_size, numel(lms.b_))
                u_index = u_size - ii;
                x_hat = x_hat + lms.u_.get(u_index) * lms.b_(ii);
            end
        end
        
        % Update the coefficients given a prior prediction and the
        % corresponding measurement.
        function Update(lms, x, x_hat)
            error = x_hat - x;
            
            % Get sizes of histories.
            x_size = lms.x_.size();
            u_size = lms.u_.size();
            
            % Gradient descent on 'a'.
            for ii = 1:numel(lms.a_)
                x_index = x_size - ii;
                if x_index < 0
                    break
                end
                
                derivative = error * lms.x_.get(x_index);
                lms.a_(ii) = lms.a_(ii) - lms.alpha_ * derivative;
            end
            
            % Gradient descent on 'a'.
            for ii = 1:numel(lms.b_)
                u_index = u_size - ii;
                if u_index < 0
                    break
                end
                derivative = error * lms.u_.get(u_index);
                lms.b_(ii) = lms.b_(ii) - lms.beta_ * derivative;
            end
            
            % Add measurement to the list.
            lms.x_.add(x);
        end        
    end
end

File: Revision_Spring2017/python/adaptive_filter.py
################################################################################
#
# Author: David Fridovich-Keil ( dfk@eecs.berkeley.edu )
# File: adaptive_filter.py
#
# Adaptive filter implementation.
#
################################################################################

import numpy as np
from collections import deque

class AdaptiveFilter:

    def __init__(self, p, q, alpha, beta):
        """ Constructor."""

        self.p_ = p # Number of autoregressive coefficients (on past latencies).
        self.q_ = q # Number of moving average coefficients (on past controls).
        self.alpha_ = alpha # Learning rate for AR coefficients.
        self.beta_ = beta # Learning rate for MA coefficients.

        # Initialize coefficients with all zeros.
        self.a_ = np.ones(p)
        self.b_ = np.ones(q)

        # History stored as a deque.
        self.l_ = deque(maxlen=p)
        self.r_ = deque(maxlen=q)

    def Predict(self, last_r=0.0, restore=False):
        """
        Pass in the next control input 'r' and predict next latency 'l'.
        If 'restore' flag is set, then does not change state. Otherwise,
        remembers thsi control input 'last_r' and potentially forgets oldest
        control.
        """
        l_hat = 0.0

        # Add control to the deque. Most recent values are on the left.
        if len(self.r_) > 0:
            oldest_r = self.r_[-1]
        self.r_.appendleft(last_r)

        # Convolve 'a' with latency history.
        for ii, l in enumerate(self.l_):
            l_hat += l * self.a_[ii]

        # Convolve 'b' with control history.
        for ii, r in enumerate(self.r_):
            l_hat += r * self.b_[ii]

        # Maybe restore to original state.
        if (restore):
            self.r_.popleft()
            if len(self.r_) == self.q_:
                self.r_.append(oldest_r)

        return l_hat

    def Update(self, l_meas, l_hat):
        """ Update 'a' and 'b' given a measurement and prediction. """
        error = l_hat - l_meas

        # Precompute the L2 norm of both latency and control histories.
        # NOTE: If this is a bottleneck, it can be sped up significantly.
        l_norm = 0.0
        for l in self.l_:
            l_norm += l * l

        r_norm = 0.0
        for r in self.r_:
            r_norm += r * r

        total_norm = l_norm + r_norm

        # Gradient descent on 'a'
        if total_norm > 1e-16:
            for ii, l in enumerate(self.l_):
                self.a_[ii] -= self.alpha_ * error * l / total_norm

        # Gradient descent on 'a'.
        if total_norm > 1e-16:
            for ii, r in enumerate(self.r_):
                self.b_[ii] -= self.beta_ * error * r / total_norm

        # Add new measurement to the history.
        self.l_.appendleft(l_meas)
